In [25]:

import math
import numpy as np
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity

In [26]:
df_bakery = pd.read_csv(r"C:\Users\stuar\Documents\College\Data Mining\Assignment\75000-out2-binary.csv")

In [27]:
df_bakery.head()

Transaction Number  Chocolate Cake  Lemon Cake  Casino Cake  Opera Cake  \
0                   1               0           0            0           0   
1                   2               0           0            0           0   
2                   3               0           0            0           1   
3                   4               0           0            0           0   
4                   5               0           0            0           0   

   Strawberry Cake  Truffle Cake  Chocolate Eclair  Coffee Eclair  \
0                0             0                 0              0   
1                0             0                 0              1   
2                0             0                 0              0   
3                0             1                 0              0   
4                0             0                 1              0   

   Vanilla Eclair  ...  Lemon Lemonade  Raspberry Lemonade  Orange Juice  \
0               0  ...               0                   0             0   
1               0  ...               0                   0             0   
2               0  ...               0                   0             1   
3               0  ...               0                   0             0   
4               0  ...               0                   0             1   

   Green Tea  Bottled Water  Hot Coffee  Chocolate Coffee  \
0          0              0           0                 0   
1          0              0           1                 0   
2          0              0           0                 0   
3          0              0           0                 0   
4          0              0           0                 0   

   Vanilla Frappuccino  Cherry Soda  Single Espresso  
0                    0            0                0  
1                    0            0                0  
2                    0            0                0  
3                    1            0                0  
4                    0            0                0  

[5 rows x 51 columns]

In [28]:
product_1 = 'Chocolate Coffee'
product_2 = 'Chocolate Cake'

select_bakery = df_bakery[[product_1, product_2]]

In [10]:
count_1p = select_bakery[product_1] == 1
count_2p = select_bakery[product_2] == 1

In [11]:
f_bake = select_bakery.groupby([count_1p, count_2p]).count()
print(f_bake)

                                 Chocolate Coffee  Chocolate Cake
Chocolate Coffee Chocolate Cake                                  
False            False                      65802           65802
                 True                        2962            2962
True             False                       2933            2933
                 True                        3303            3303


In [12]:
f00 = f_bake[product_1][0][0]
f01 = f_bake[product_1][0][1]
f10 = f_bake[product_2][1][0]
f11 = f_bake[product_2][1][1]

print("f00 -> ",f00)
print("f01 -> ",f01)
print("f10 -> ",f10)
print("f11 -> ",f11)

f00 ->  65802
f01 ->  2962
f10 ->  2933
f11 ->  3303


In [13]:
f1plus = f11 + f10
fplus1 = f11 + f01
f0plus = f01 + f00
fplus0 = f10 + f00

print("f1+ -> ",f1plus)
print("f+1 -> ",fplus1)
print("f0+ -> ",f0plus)
print("f+0 -> ",fplus0)

f1+ ->  6236
f+1 ->  6265
f0+ ->  68764
f+0 ->  68735


In [17]:
n_total = f1plus + f0plus
print("Total -> ",n_total)

n_total = fplus1 + fplus0
print("Total -> ",n_total)

Total ->  75000
Total ->  75000


In [20]:
print("\t","Coco","\t\t","!Coco","\t")
print()
print("Cake", "\t",f11,"\t",f10,"\t",f1plus)
print()
print("!Cake","\t",f01,"\t",f00,"\t",f0plus)
print()
print("\t", fplus1, "\t", fplus0, "\t", n_total)

	 Coco 		 !Coco 	

Cake 	 3303 	 2933 	 6236

!Cake 	 2962 	 65802 	 68764

	 6265 	 68735 	 75000


In [22]:
s = f11 / n_total
print("Support -> ",s)

c = f11 / f1plus
print("Count   -> ",c)

Support ->  0.04404
Count   ->  0.5296664528543938


In [23]:
l = c / (fplus1 / n_total)
print("Lift            -> ",l)

I = n_total*f11 / (f1plus * fplus1)
print("Interest Factor -> ",I)

Lift            ->  6.340779563300804
Interest Factor ->  6.340779563300805


In [37]:
s_a = f1plus / n_total
s_b = fplus1 / n_total
s_ab = f11 / n_total

c_ab = s_ab / s_a
c_ba = s_ab / s_b

IS = s_ab / (math.sqrt(s_a * s_b))
print("IS measures(using support) ->",IS)

IS = math.sqrt(c_ab * c_ba)
print("IS measures (using count)  ->",IS)

IS measures(using support) -> 0.5284391468918322
IS measures (using count)  -> 0.5284391468918322


In [24]:
correlation = ((f11*f00)-(f10*f01))/(math.sqrt(f1plus*fplus1*f0plus*fplus0))
print("Binary Correlation Coefficient->",correlation)

Binary Correlation Coefficient-> 0.4855664925278768


#### The binary correlation is 0.4856. 
#### These 2 items are Assymmetric Binary Variables.

#### s({Chocolate Cake}) = 0.0835
#### s({Chocolate Coffee}) = 0.08314
#### s({Chocolate Cake},{Chocolate Coffee}) = 0.4404

#### Yes, The correlation coefficient is the same value for both the associations rules: {Chocolate Coffee} -> {Chocolate Cake} and {Chocolate Cake} -> {Chocolate Coffee} because the correlation coefficient is invariant